agent/agent.go

```diff
@@ -157,247 +150,51 @@ func NewAgent(config *Config) *Agent {
-				logger.Debugf("error reading message from tcp socket %s", err.Error())
-				return
-			}
-		}
-
-		if match := pingRe.Match(messageBuffer.Bytes()); match {
-			logger.Debug("tcp socket received ping")
-			_, err = c.Write([]byte("pong"))
-			if err != nil {
-				logger.WithError(err).Error("could not write response to tcp socket")
-			}
-			return
-		}
-		// Check our received data for valid JSON. If we get invalid JSON at this point,
-		// read again from client, add any new message to the buffer, and parse
-		// again.
-		var event types.Event
-		var result v1.CheckResult
-		if err = json.Unmarshal(messageBuffer.Bytes(), &result); err != nil {
-			continue
-		}
-
-		if err = translateToEvent(a, result, &event); err != nil {
-			logger.WithError(err).Error("1.x returns \"invalid\"")
-			return
-		}
-
-		// Prepare the event by mutating it as required so it passes validation
-		if err = prepareEvent(a, &event); err != nil {
-			logger.WithError(err).Error("invalid event")
-			return
-		}
-
-		// At this point, should receive valid JSON, so send it along to the
-		// message sender.
-		payload, err := json.Marshal(event)
-		if err != nil {
-			logger.WithError(err).Error("could not marshal json payload")
-			return
-		}
-
-		a.sendMessage(transport.MessageTypeEvent, payload)
-		_, _ = c.Write([]byte("ok"))
-		return
-	}
-	_, _ = c.Write([]byte("invalid"))
-}
-
-// If the socket receives a message containing whitespace and the
-// string "ping", it will ignore it.
-//
-// The socket assumes all other messages will contain a single,
-// complete, JSON hash. The hash must be a valid JSON check result.
-// Deserialization failures will be logged at the ERROR level by the
-// Sensu agent, but the sender of the invalid data will not be
-// notified.
-func (a *Agent) handleUDPMessages(c net.PacketConn) {
-	var buf [1500]byte
-
-	go func() {
-		<-a.stopping
-		if err := c.Close(); err != nil {
-			logger.Debug(err)
-		}
-		a.wg.Done()
-	}()
-	// Read everything sent from the connection to the message buffer. Any error
-	// will return. If the buffer is zero bytes, close the connection and return.
-	for {
-		bytesRead, _, err := c.ReadFrom(buf[0:])
-		select {
-		case <-a.stopping:
-			logger.Debug("UDP listener stopped")
-			return
-		default:
-			if err != nil {
-				logger.WithError(err).Error("Error reading from UDP socket")
-				if err := c.Close(); err != nil {
-					logger.Debug(err)
+				// Now, we must attempt to reconnect to the backend, with exponential
+				// backoff
+				backoff := retry.ExponentialBackoff{
+					InitialDelayInterval: 500 * time.Millisecond,
+					MaxDelayInterval:     10 * time.Second,
+					MaxRetryAttempts:     0, // Unlimited attempts
+					Multiplier:           1.5,
 				}
-				return
-			} else if bytesRead == 0 {
-				if err := c.Close(); err != nil {
-					logger.Debug(err)
+				if err := backoff.Retry(func(retry int) (bool, error) {
+					logger.Debugf("reconnection attempt #%d", retry)
```

Comment:
Probably worth guarding this with 
```
if retry > 0 {
}
```
to avoid producing 'retry 0' debug calls in the common case.

Reply:
Good catch!